share/doc/wake/tutorial.md

```diff
@@ -23,691 +32,1166 @@ The output of wake run on `-x 'expression'` is the result of evaluating that exp
+files in the same directory and pass them as legal inputs to gcc -- the
+keyword `here` expands to the directory of the `.wake` file.  The second
+argument to `sources` is a regular expression to select which files to
+return. We've used ``` `` ```s here which define regular expression literals with the [standard syntax](https://github.com/google/re2/wiki/Syntax).
+In addition, the parser verifies that regular expression literals are legal.
 
-Since the `source` will return a `BadPath`, the job will propagate the failure:
+Note that "source files" are those files tracked by git. Wake will never
+return built files from a call to `sources`, helping repeatability. We can see
+this through the fact that the `info.h` file we generated is not listed in the
+visible list, despite the regex supposedly matching all `.h` files in the
+current directory. If we did want to use it, we'd need to reference the `Path`
+returned when it is created (or retrieved from the cache):
+
+```wake
+def infoResult =
+    infoH Nil
+def compile =
+    require Pass info = infoResult
+    def visible =
+        info, headers
+    compileC variant ("-I.", Nil) visible
+```
 
-    BadPath (Error "file.c: not a source file" Nil), Nil
+In a classic Makefile it would be considered bad form to list all header
+files as dependencies for all cpp files.  That's because make would
+recompile every cpp file whenever any header file changes.  In wake, we
+don't have this problem.  Wake monitors jobs to see which files they
+actually used and remembers this for later builds.  Therefore, it's best in
+wake to err on the side of caution (and convenience) by just listing all the
+headers in directories that are interesting to the cpp files.
 
-This means that when passing the outputs from one job to another,
-you don't have to worry about jobs failing, wake will handle it for you!
+```console
+$ wake -o main.native-cpp11-release.o
+...
+Inputs:
+  d2a7bfde help.h
+  9d220741 main.cpp
+Outputs:
+  810e17cb main.native-cpp11-release.o
+```
+
+For this file, wake recorded that it needed both `main.cpp` and `help.h`.
+
+```console
+$ wake -o help.native-cpp11-release.o
+...
+Inputs:
+  5b8beead help.cpp
+Outputs:
+  13350319 help.native-cpp11-release.o
+```
 
-You may have noticed that the `BadPath` above contains a type called `Error`.
-`Error` is a type that contains a `String` "cause", and a `List` of `Strings` stack trace.
-You'll notice the cause is `"file.c: not a source file"`, but the stack is just `Nil`.
-Wake does not actually maintain a call stack like traditional languages,
-so by default `Errors` will contain an empty `List` for the stack.
-If you run wake with `-d` (or `--debug`), it will simulate a stack:
+For this file, wake recorded that it only needed `help.cpp`, despite
+`help.h` being a legal input.
 
-    wake -dx 'job (which "gcc", "file.c", Nil) (source "file.c", Nil) | getJobOutputs'
+## Publish/Subscribe
```

Comment:
This might be better elsewhere rather than sandwiched between visible files and procedural output-to-target.